[DTensor] Fix empty shard global-offset calculation

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150862

`compute_local_shape_and_global_offset` util computes the local shape of
a particular shard of a DTensor, and the global offset (which describes
how the shard fits into the global tensor).

When the tensor dim does not evenly divide into the mesh dim, uneven
sharding occurs.  In some cases, uneven sharding results in an empty
shard.

e.g.
   tensor dim size: 4096
   mesh dim size: 30
   ranks 0..27 have local size 18
   rank 28 has local size 8
   rank 29 has local size 0 <--- empty shard

The global offset for an empty shard was previously undefined and
returned values that were computed based on logic that assumes no empty
shards.  This caused DCP to fail to save a checkpoint, becuase
deduplication logic could 'throw away' real (non-empty) shards thinking
they were duplicates of zero-sized shards with the same offset.

Now, we define the global offset of an empty shard to be the dim-size,
which is out of bounds of the tensor and can't overlap with any
non-empty shards.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

wconstab committed:
WIP

[ghstack-poisoned]

wconstab committed:
Update on "WIP test a corner case"




cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "WIP test a corner case"




cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Fix empty shard global-offset calculation"


`compute_local_shape_and_global_offset` util computes the local shape of
a particular shard of a DTensor, and the global offset (which describes
how the shard fits into the global tensor).

When the tensor dim does not evenly divide into the mesh dim, uneven
sharding occurs.  In some cases, uneven sharding results in an empty
shard.

e.g.
   tensor dim size: 4096
   mesh dim size: 30
   ranks 0..27 have local size 18
   rank 28 has local size 8
   rank 29 has local size 0 <--- empty shard

The global offset for an empty shard was previously undefined and
returned values that were computed based on logic that assumes no empty
shards.  This caused DCP to fail to save a checkpoint, becuase
deduplication logic could 'throw away' real (non-empty) shards thinking
they were duplicates of zero-sized shards with the same offset.

Now, we define the global offset of an empty shard to be the dim-size,
which is out of bounds of the tensor and can't overlap with any
non-empty shards.

[ghstack-poisoned]

teja-rao reviewed (commented):
looks good to me

teja-rao commented on torch/distributed/tensor/_utils.py:
```diff
@@ -110,18 +110,29 @@ def compute_local_shape_and_global_offset(
+
 
-    my_coordinate = mesh.get_coordinate()
+# accept 'plain data types' to enable simpler unit testing without creating device mesh
```
thanks!

wconstab commented:
@XilunWu was advocating for a change to https://github.com/pytorch/pytorch/pull/150862 which is to return an empty tuple (,) as the global offset in the size-zero shard case.

The summary of our discussion was:

Returning 'dimsize' for the offset value of zero-sized shard dimensions is at least explainable, and has the bonus property of avoiding a current DCP bug in potentially dropping a valid shard and keeping an empty shard thinking wrongly that they are duplicates.  However it is still possibly misleading, because we return an integer value that could be interpreted as an offset when in fact there is no offset.

The (,) approach seems even more clear, there is no wrong number to look at, but it is also costlier in that you need to deal with any BC issue with callers that expect to get a non-empty tuple.  Like anyone who does zip(mesh_sizes, offsets) will have to rewrite their code..

cc @teja-rao

XilunWu commented on test/distributed/tensor/test_utils.py:
```diff
@@ -102,6 +103,37 @@ def test_explicit_order_placements(self):
                 )
 
+    def test_compute_local_shape_and_global_offset_uneven(self):
+        # This case is not only 'uneven' bug also has an empty shard
```
need clean up the test:
- remove print
- change the if-else block to
```
exp_local_shape = ...
local_shape = ...
self.assertEqual(local_shape, exp_local_shape)
self.assertEqual(global_offset, exp_global_offset)
```

wconstab replied:
cleaned up.

XilunWu reviewed (approved):
LGTM. Stamp to unblock

wconstab commented on torch/distributed/tensor/_utils.py:
```diff
@@ -110,18 +110,29 @@ def compute_local_shape_and_global_offset(
+    return _compute_local_shape_and_global_offset(
+        global_shape, mesh.shape, mesh.get_coordinate(), placements
+    )
+
```
yea I think having a FakeDeviceMesh would be great. I can put a todo there.

wconstab committed:
Update on "[DTensor] Fix empty shard global-offset calculation"


`compute_local_shape_and_global_offset` util computes the local shape of
a particular shard of a DTensor, and the global offset (which describes
how the shard fits into the global tensor).

When the tensor dim does not evenly divide into the mesh dim, uneven
sharding occurs.  In some cases, uneven sharding results in an empty
shard.

e.g.
   tensor dim size: 4096
   mesh dim size: 30
   ranks 0..27 have local size 18
   rank 28 has local size 8
   rank 29 has local size 0 <--- empty shard

The global offset for an empty shard was previously undefined and
returned values that were computed based on logic that assumes no empty
shards.  This caused DCP to fail to save a checkpoint, becuase
deduplication logic could 'throw away' real (non-empty) shards thinking
they were duplicates of zero-sized shards with the same offset.

Now, we define the global offset of an empty shard to be the dim-size,
which is out of bounds of the tensor and can't overlap with any
non-empty shards.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -i
